src/main/cc/common_cpp/protobuf_util/riegeli_io.h

```diff
@@ -0,0 +1,101 @@
+// Copyright 2021 The Cross-Media Measurement Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef SRC_MAIN_CC_COMMON_CPP_PROTOBUF_UTIL_RIEGELI_IO_H_
+#define SRC_MAIN_CC_COMMON_CPP_PROTOBUF_UTIL_RIEGELI_IO_H_
+
+#include <fcntl.h>
+#include <stdio.h>
+
+#include <vector>
+
+#include "absl/status/status.h"
+#include "absl/strings/str_cat.h"
+#include "absl/strings/string_view.h"
+#include "riegeli/bytes/fd_reader.h"
+#include "riegeli/bytes/fd_writer.h"
+#include "riegeli/records/record_reader.h"
+#include "riegeli/records/record_writer.h"
+
+namespace wfa {
+
+using ::riegeli::FdReader;
+using ::riegeli::FdWriter;
+using ::riegeli::RecordReader;
+using ::riegeli::RecordWriter;
+
+// Read a list of protobuf messages to @messages. The input messages are from
+// the file with @path.
+// Return error status if the reading fails.
+// The details about the format of the input file can be found in
+// https://github.com/google/riegeli/blob/master/doc/riegeli_records_file_format.md
+template <typename Message>
+absl::Status ReadRiegeliFile(absl::string_view path,
+                             std::vector<Message>& messages) {
+  RecordReader reader(FdReader(path, O_RDONLY));
+  messages.clear();
+  while (true) {
+    Message message;
+    if (!reader.ReadRecord(message)) {
+      if (reader.healthy()) {
+        // EOF
+        break;
+      } else {
+        return absl::InvalidArgumentError(
+            absl::StrCat("Unable to read message from file: ", path));
+      }
+    }
+    messages.emplace_back(message);
+  }
+  if (!reader.Close()) {
+    return reader.status();
+  }
+  return absl::OkStatus();
+}
+
+// Write the list of protobuf messages in @messages to the file with @path.
+// Return error status if the writing fails.
+// The details about the format of the output file can be found in
+// https://github.com/google/riegeli/blob/master/doc/riegeli_records_file_format.md
+template <typename Message>
+absl::Status WriteRiegeliFile(absl::string_view path,
+                              const std::vector<Message>& messages) {
+  // Create the file at @path if not exists.
+  FILE* file = fopen(path.data(), "r");
+  if (!file) {
+    file = fopen(path.data(), "w");
+    if (!file) {
+      return absl::InternalError(absl::StrCat("Unable to create file: ", path));
+    }
+  }
+  if (fclose(file) != 0) {
+    return absl::InternalError(absl::StrCat("Unable to close file: ", path));
+  }
```

Comment:
This should not be needed. Instead, pass `O_WRONLY | O_CREAT | O_TRUNC` to `FdWriter`.